govcd/cse.go

```diff
@@ -0,0 +1,379 @@
+		}
+		// The OVA can be used if the TKG version is higher than the actual and the Kubernetes version is at most 1 minor higher.
+		if targetVersions.compareTkgVersion(cluster.TkgVersion.String()) == 1 && targetVersions.kubernetesVersionIsOneMinorHigher(cluster.KubernetesVersion.String()) {
+			cluster.supportedUpgrades = append(cluster.supportedUpgrades, vAppTemplate.VAppTemplate)
+		}
+	}
+	return cluster.supportedUpgrades, nil
+}
+
+// UpgradeCluster executes an update on the receiver cluster to upgrade the Kubernetes template of the cluster.
+// If the cluster is not upgradeable or the OVA is incorrect, this method will return an error.
+// If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+func (cluster *CseKubernetesCluster) UpgradeCluster(kubernetesTemplateOvaId string, refresh bool) error {
+	return cluster.Update(CseClusterUpdateInput{
+		KubernetesTemplateOvaId: &kubernetesTemplateOvaId,
+	}, refresh)
+}
+
+// SetNodeHealthCheck executes an update on the receiver cluster to enable or disable the machine health check capabilities.
+// If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+func (cluster *CseKubernetesCluster) SetNodeHealthCheck(healthCheckEnabled bool, refresh bool) error {
+	return cluster.Update(CseClusterUpdateInput{
+		NodeHealthCheck: &healthCheckEnabled,
+	}, refresh)
+}
+
+// SetAutoRepairOnErrors executes an update on the receiver cluster to change the flag that controls the auto-repair
+// capabilities of CSE. If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+// NOTE: This method can only be used in CSE versions < 4.1.1
+func (cluster *CseKubernetesCluster) SetAutoRepairOnErrors(autoRepairOnErrors bool, refresh bool) error {
+	return cluster.Update(CseClusterUpdateInput{
+		AutoRepairOnErrors: &autoRepairOnErrors,
+	}, refresh)
+}
+
+// Update executes an update on the receiver CSE Kubernetes Cluster on any of the allowed parameters defined in the input type.
+// If refresh=true, it retrieves the latest state of the cluster from VCD before updating.
+func (cluster *CseKubernetesCluster) Update(input CseClusterUpdateInput, refresh bool) error {
+	if refresh {
+		err := cluster.Refresh()
+		if err != nil {
+			return err
+		}
+	}
+
+	if cluster.State == "" {
+		return fmt.Errorf("can't update a Kubernetes cluster that does not have any state")
+	}
+	if cluster.State != "provisioned" {
+		return fmt.Errorf("can't update a Kubernetes cluster that is not in 'provisioned' state, as it is in '%s'", cluster.capvcdType.Status.VcdKe.State)
+	}
+
+	if input.AutoRepairOnErrors != nil {
+		// Since CSE 4.1.1, the AutoRepairOnError toggle can't be modified and is turned off
+		// automatically by the CSE Server.
+		v411, err := semver.NewVersion("4.1.1")
+		if err != nil {
+			return fmt.Errorf("can't update the  'Auto Repair on Errors' flag: %s", err)
+		}
```

Comment:
This error looks to incorrect. Is it simply an error that would happen if `semver.NewVersion("4.1.1")` couldn't be parsed?

Reply:
Yeah, in reality it is a bit dumb, as the version is hardcoded and should never fail, but I added the error for completeness. Will change it to
```
if err != nil {
    return err
}
```